### ISSUE TITLE
Restart-BCContainer with renewBindings fails because container is not running?

### ISSUE THREAD

**Issue by epernst** [microsoft/navcontainerhelper#3071]
Sometimes, but not always, when we use Restart-BcContainer $ContainerName -RenewBindings to apply our Let's Encrypt SSL certificates to containers, then it fails with: "**Error response from daemon: Container xxxx is not running**" (xxxx = containerid).

The container status is "Exited (1073807364) x minutes ago", so it is correct that it doesn't run. But it's never running, before we apply the SSL, and most of the times, it work. 

Would I have run a Start-BcContainer before calling Restart-BcContainer? Even if Start-BcContainer doesn't have the -RenewBindings param?

(Using BCCH version 5.03)

**Comment by freddydk** [comment]
Restart-BcContainer should not have a problem with starting a stopped container.
Could you show me the full output of the command? (from command to error)

**Comment by epernst** [comment]
Azure DevOps logs comes here:
2023-05-26T15:22:20.8892364Z ##[debug]Evaluating condition for step: 'Configure container'
2023-05-26T15:22:20.8894131Z ##[debug]Evaluating: eq(and(succeeded(), eq(variables['vmExists'], 'false')), True)
2023-05-26T15:22:20.8894373Z ##[debug]Evaluating eq:
2023-05-26T15:22:20.8894545Z ##[debug]..Evaluating and:
2023-05-26T15:22:20.8894843Z ##[debug]....Evaluating succeeded:
2023-05-26T15:22:20.8895296Z ##[debug]....=> True
2023-05-26T15:22:20.8895483Z ##[debug]....Evaluating eq:
2023-05-26T15:22:20.8895674Z ##[debug]......Evaluating indexer:
2023-05-26T15:22:20.8896101Z ##[debug]........Evaluating variables:
2023-05-26T15:22:20.8896481Z ##[debug]........=> Object
2023-05-26T15:22:20.8896695Z ##[debug]........Evaluating String:
2023-05-26T15:22:20.8896907Z ##[debug]........=> 'vmExists'
2023-05-26T15:22:20.8897199Z ##[debug]......=> 'false'
2023-05-26T15:22:20.8897405Z ##[debug]......Evaluating String:
2023-05-26T15:22:20.8897593Z ##[debug]......=> 'false'
2023-05-26T15:22:20.8897872Z ##[debug]....=> True
2023-05-26T15:22:20.8898083Z ##[debug]..=> True
2023-05-26T15:22:20.8898238Z ##[debug]..Evaluating Boolean:
2023-05-26T15:22:20.8898390Z ##[debug]..=> True
2023-05-26T15:22:20.8898525Z ##[debug]=> True
2023-05-26T15:22:20.8898952Z ##[debug]Expanded: eq(and(True, eq('false', 'false')), True)
2023-05-26T15:22:20.8899206Z ##[debug]Result: True
2023-05-26T15:22:20.8899488Z ##[section]Starting: Configure container
2023-05-26T15:22:20.9056960Z ==============================================================================
2023-05-26T15:22:20.9057143Z Task         : PowerShell
2023-05-26T15:22:20.9057217Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2023-05-26T15:22:20.9057326Z Version      : 2.220.0
2023-05-26T15:22:20.9057397Z Author       : Microsoft Corporation
2023-05-26T15:22:20.9057495Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2023-05-26T15:22:20.9057616Z ==============================================================================
2023-05-26T15:22:21.6625226Z ##[debug]VstsTaskSdk 0.11.0 commit 7ff27a3e0bdd6f7b06690ae5f5b63cb84d0f23f4
2023-05-26T15:22:21.8857880Z ##[debug]Entering C:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.220.0\powershell.ps1.
2023-05-26T15:22:21.8971125Z ##[debug]Loading resource strings from: C:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.220.0\task.json
2023-05-26T15:22:21.9387691Z ##[debug]Loaded 13 strings.
2023-05-26T15:22:21.9459845Z ##[debug]SYSTEM_CULTURE: 'en-US'
2023-05-26T15:22:21.9502553Z ##[debug]Loading resource strings from: C:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.220.0\Strings\resources.resjson\en-US\resources.resjson
2023-05-26T15:22:21.9827896Z ##[debug]Loaded 13 strings.
2023-05-26T15:22:22.0124970Z ##[debug]INPUT_ERRORACTIONPREFERENCE: 'stop'
2023-05-26T15:22:22.0210925Z ##[debug]INPUT_WARNINGPREFERENCE: 'default'
2023-05-26T15:22:22.0261771Z ##[debug]INPUT_INFORMATIONPREFERENCE: 'default'
2023-05-26T15:22:22.0309431Z ##[debug]INPUT_VERBOSEPREFERENCE: 'default'
2023-05-26T15:22:22.0355348Z ##[debug]INPUT_DEBUGPREFERENCE: 'default'
2023-05-26T15:22:22.0402361Z ##[debug]INPUT_PROGRESSPREFERENCE: 'silentlyContinue'
2023-05-26T15:22:22.0447556Z ##[debug]INPUT_SHOWWARNINGS: 'false'
2023-05-26T15:22:22.0495188Z ##[debug] Converted to bool: False
2023-05-26T15:22:22.0541989Z ##[debug]INPUT_FAILONSTDERR: 'true'
2023-05-26T15:22:22.0577540Z ##[debug] Converted to bool: True
2023-05-26T15:22:22.0622250Z ##[debug]INPUT_IGNORELASTEXITCODE: 'false'
2023-05-26T15:22:22.0657990Z ##[debug] Converted to bool: False
2023-05-26T15:22:22.0724520Z ##[debug]INPUT_PWSH: 'false'
2023-05-26T15:22:22.0742941Z ##[debug] Converted to bool: False
2023-05-26T15:22:22.0789565Z ##[debug]INPUT_WORKINGDIRECTORY: 'C:\a\2\s'
2023-05-26T15:22:22.0954740Z ##[debug]Asserting container path exists: 'C:\a\2\s'
2023-05-26T15:22:22.1005598Z ##[debug]INPUT_TARGETTYPE: 'inline'
2023-05-26T15:22:22.1668417Z ##[debug]INPUT_SCRIPT: 'Import-Module -FullyQualifiedName C:\a\2/ReleasePipeline/DevOpsRelease/DBC.DevOps.ReleaseTools/DBC.DevOps.ReleaseTools.psm1 -DisableNameChecking -Force 
2023-05-26T15:22:22.1690119Z ##[debug]$RepoSettings = Get-Content "C:\a\2\s/setup/settings.json" | Out-String | ConvertFrom-Json 
2023-05-26T15:22:22.1710327Z ##[debug]$Credential = [PSCredential]::new("$($env:VMUSERNAME)", (ConvertTo-SecureString -String "$($env:VMPASSWORD)" -AsPlainText -Force))
2023-05-26T15:22:22.1730907Z ##[debug]$Session = New-WinRmSession -ConnectionUri https://xxxxxx.westeurope.cloudapp.azure.com:5986 -Credential $Credential -ImportDevOpsBuild
2023-05-26T15:22:22.1752222Z ##[debug]$Params = @{
2023-05-26T15:22:22.1773420Z ##[debug]    containerName              = "xxxx-platform"
2023-05-26T15:22:22.1793997Z ##[debug]    ContainerHelperScripts     = "C:\DBC.DevOps\ContainerHelperScripts"
2023-05-26T15:22:22.1814886Z ##[debug]    publicDnsName              = "xxxxxx.westeurope.cloudapp.azure.com"
2023-05-26T15:22:22.1842677Z ##[debug]    ContactEMailForLetsEncrypt = xx
2023-05-26T15:22:22.1871571Z ##[debug]    CertificatePfxPassword     = xx
2023-05-26T15:22:22.1898348Z ##[debug]    licenseFile                = $RepoSettings.ContainerParams.licenseFile      
2023-05-26T15:22:22.1927821Z ##[debug]} 
2023-05-26T15:22:22.1953860Z ##[debug]Invoke-Command -Session $Session { Param( [HashTable] $Params)      
2023-05-26T15:22:22.1977586Z ##[debug]    Set-ContainerCertificate @Params
2023-05-26T15:22:22.1998996Z ##[debug]} -Argumentlist $Params
2023-05-26T15:22:22.2025374Z ##[debug]Remove-RemoteAdminSession -Session $Session'
2023-05-26T15:22:22.2070366Z ##[debug]INPUT_RUNSCRIPTINSEPARATESCOPE: 'false'
2023-05-26T15:22:22.2105204Z ##[debug] Converted to bool: False
2023-05-26T15:22:22.2174181Z Generating script.
2023-05-26T15:22:22.2413735Z ##[debug]AGENT_VERSION: '3.220.2'
2023-05-26T15:22:22.2514478Z ##[debug]AGENT_TEMPDIRECTORY: 'C:\a\_temp'
2023-05-26T15:22:22.2546199Z ##[debug]Asserting container path exists: 'C:\a\_temp'
2023-05-26T15:22:22.2738605Z ##[debug]Asserting leaf path exists: 'C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe'
2023-05-26T15:22:22.2759283Z ========================== Starting Command Output ===========================
2023-05-26T15:22:22.2900070Z ##[debug]Entering Invoke-VstsTool.
2023-05-26T15:22:22.3027792Z ##[debug] Arguments: '-NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\a\_temp\bf61ed90-b4bd-4edb-8995-26641d616859.ps1'"'
2023-05-26T15:22:22.3076761Z ##[debug] FileName: 'C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe'
2023-05-26T15:22:22.3111348Z ##[debug] WorkingDirectory: 'C:\a\2\s'
2023-05-26T15:22:22.3186448Z ##[command]"C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\a\_temp\bf61ed90-b4bd-4edb-8995-26641d616859.ps1'"
2023-05-26T15:22:51.8313499Z BcContainerHelper version 5.0.3
2023-05-26T15:22:51.8372783Z DEBUG: 
2023-05-26T15:22:51.8387750Z  using System.Net;
2023-05-26T15:22:51.8394995Z 
2023-05-26T15:22:51.8408780Z  
2023-05-26T15:22:51.8414696Z 
2023-05-26T15:22:51.8428432Z  public class TimeoutWebClient : WebClient
2023-05-26T15:22:51.8433317Z 
2023-05-26T15:22:51.8446002Z  {
2023-05-26T15:22:51.8453276Z 
2023-05-26T15:22:51.8468341Z         int theTimeout;
2023-05-26T15:22:51.8473352Z 
2023-05-26T15:22:51.8479112Z 
2023-05-26T15:22:51.8485180Z 
2023-05-26T15:22:51.8498646Z         public TimeoutWebClient(int timeout)
2023-05-26T15:22:51.8503497Z 
2023-05-26T15:22:51.8517306Z         {
2023-05-26T15:22:51.8523237Z 
2023-05-26T15:22:51.8535685Z             theTimeout = timeout;
2023-05-26T15:22:51.8542673Z 
2023-05-26T15:22:51.8555192Z         }
2023-05-26T15:22:51.8560993Z 
2023-05-26T15:22:51.8567677Z 
2023-05-26T15:22:51.8573880Z 
2023-05-26T15:22:51.8586397Z   protected override WebRequest GetWebRequest(System.Uri address)
2023-05-26T15:22:51.8593163Z 
2023-05-26T15:22:51.8605309Z   {
2023-05-26T15:22:51.8611124Z 
2023-05-26T15:22:51.8624831Z    WebRequest request = base.GetWebRequest(address);
2023-05-26T15:22:51.8629354Z 
2023-05-26T15:22:51.8642683Z    if (request != null)
2023-05-26T15:22:51.8647083Z 
2023-05-26T15:22:51.8660535Z    {
2023-05-26T15:22:51.8666435Z 
2023-05-26T15:22:51.8680487Z     request.Timeout = theTimeout;
2023-05-26T15:22:51.8686794Z 
2023-05-26T15:22:51.8699972Z    }
2023-05-26T15:22:51.8706764Z 
2023-05-26T15:22:51.8720143Z    return request;
2023-05-26T15:22:51.8725141Z 
2023-05-26T15:22:51.8738556Z   }
2023-05-26T15:22:51.8743373Z 
2023-05-26T15:22:51.8756078Z   }
2023-05-26T15:22:51.8769798Z BC.HelperFunctions emits usage statistics telemetry to Microsoft
2023-05-26T15:22:59.8133336Z ##[section]Configuring Container with Let's Encrypt SSL Certificate
2023-05-26T15:23:09.5971524Z 
2023-05-26T15:23:09.5981540Z 
2023-05-26T15:23:09.5999634Z name       : xxxx-platform
2023-05-26T15:23:09.6017109Z id         : c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4
2023-05-26T15:23:09.6034696Z image      : xxxx/onprem:14.44.49619.0-w1
2023-05-26T15:23:09.6048699Z createdat  : 2023-05-26 11:25:41 +0000 GMT
2023-05-26T15:23:09.6063229Z runningfor : 4 hours ago
2023-05-26T15:23:09.6079531Z size       : 0B
2023-05-26T15:23:09.6094032Z status     : Exited (1073807364) 4 hours ago
2023-05-26T15:23:09.6110041Z created    : 202305261111
2023-05-26T15:23:09.6123604Z platform   : 14.0.49616.0
2023-05-26T15:23:09.6137634Z tag        : 1.0.2.14
2023-05-26T15:23:09.6151898Z version    : 14.44.49619.0
2023-05-26T15:23:09.6165649Z nav        : 
2023-05-26T15:23:09.6179411Z osversion  : 10.0.14393.5921
2023-05-26T15:23:09.6193215Z country    : w1
2023-05-26T15:23:09.6207295Z cu         : 
2023-05-26T15:23:09.6223448Z eula       : https://go.microsoft.com/fwlink/?linkid=861843
2023-05-26T15:23:09.6237276Z legal      : http://go.microsoft.com/fwlink/?LinkId=837447
2023-05-26T15:23:09.6250922Z maintainer : Dynamics SMB
2023-05-26T15:23:09.6257108Z 
2023-05-26T15:23:09.6263305Z 
2023-05-26T15:23:09.6269494Z 
2023-05-26T15:23:09.6283192Z Updating SetupVariables.ps1
2023-05-26T15:23:09.6297999Z Updating other MyScripts overrides
2023-05-26T15:23:09.6313023Z Copies C:\DBC.DevOps\ContainerHelperScripts\AdditionalSetup.ps1 to C:\ProgramData\BcContainerHelper\extensions\xxxx-platform\my
2023-05-26T15:23:09.6328057Z Copies C:\DBC.DevOps\ContainerHelperScripts\SetupCertificate.ps1 to C:\ProgramData\BcContainerHelper\extensions\xxxx-platform\my
2023-05-26T15:23:09.6342338Z Restarting container to apply settings
2023-05-26T15:23:12.5779081Z Restart-BcContainer Telemetry Correlation Id: 5b0d321a-a5fb-4a45-a217-ffbf38fa93f1
2023-05-26T15:23:12.5796827Z Attempt 1 of 5 failed: Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:12.5811988Z Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:12.5826537Z Waiting for 5 seconds before retrying...
2023-05-26T15:23:17.5790360Z Retrying...
2023-05-26T15:23:18.6765743Z Restart-BcContainer Telemetry Correlation Id: 45f10b77-b0b8-4cef-a17e-0a3e210134f2
2023-05-26T15:23:18.6781891Z Attempt 2 of 5 failed: Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:18.6796148Z Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:18.6811091Z Waiting for 5 seconds before retrying...
2023-05-26T15:23:23.7004621Z Retrying...
2023-05-26T15:23:24.6871239Z Restart-BcContainer Telemetry Correlation Id: f6a4321a-d1da-48f7-823f-e6d88d365211
2023-05-26T15:23:24.6884381Z Attempt 3 of 5 failed: Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:24.6898599Z Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:24.6914129Z Waiting for 5 seconds before retrying...
2023-05-26T15:23:29.7069675Z Retrying...
2023-05-26T15:23:30.6457959Z Restart-BcContainer Telemetry Correlation Id: 61d3d965-0fd7-4de4-bdba-077b0c0aba04
2023-05-26T15:23:30.6472608Z Attempt 4 of 5 failed: Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:30.6485676Z Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:30.6497044Z Waiting for 5 seconds before retrying...
2023-05-26T15:23:35.6675332Z Retrying...
2023-05-26T15:23:36.0564124Z Restart-BcContainer Telemetry Correlation Id: a0f906d4-090e-4164-aed8-cc9dd19c333c
2023-05-26T15:23:36.1461310Z ##[debug]STDERR: Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
2023-05-26T15:23:36.1499420Z ##[debug]STDERR: At C:\a\_temp\bf61ed90-b4bd-4edb-8995-26641d616859.ps1:16 char:1
2023-05-26T15:23:36.1525195Z ##[debug]
2023-05-26T15:23:36.1559664Z ##[debug]STDERR: + Invoke-Command -Session $Session { Param( [HashTable] $Params)
2023-05-26T15:23:36.1583824Z ##[debug]
2023-05-26T15:23:36.1619080Z ##[debug]STDERR: + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-05-26T15:23:36.1644640Z ##[debug]
2023-05-26T15:23:36.2079702Z ##[debug]STDERR:     + CategoryInfo          : NotSpecified: (Error response ... is not running:String) [], RemoteException
2023-05-26T15:23:36.2102092Z ##[debug]
2023-05-26T15:23:36.2150204Z ##[debug]STDERR:     + FullyQualifiedErrorId : NativeCommandError
2023-05-26T15:23:36.2171964Z ##[debug]
2023-05-26T15:23:36.2200502Z ##[debug]STDERR:     + PSComputerName        : xxxxxx.westeurope.cloudapp.azure.com
2023-05-26T15:23:36.2219575Z ##[debug]
2023-05-26T15:23:36.2251192Z ##[debug]STDERR:  
2023-05-26T15:23:36.2269655Z ##[debug]
2023-05-26T15:23:36.2302847Z ##[debug]Exit code: 1
2023-05-26T15:23:36.2359503Z ##[debug]Leaving Invoke-VstsTool.
2023-05-26T15:23:36.2696153Z ##[error]Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
At C:\a\_temp\bf61ed90-b4bd-4edb-8995-26641d616859.ps1:16 char:1

+ Invoke-Command -Session $Session { Param( [HashTable] $Params)

+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    + CategoryInfo          : NotSpecified: (Error response ... is not running:String) [], RemoteException

    + FullyQualifiedErrorId : NativeCommandError

    + PSComputerName        : xxxxxx.westeurope.cloudapp.azure.com
2023-05-26T15:23:36.2706458Z ##[debug]Processed: ##vso[task.logissue type=error]Error response from daemon: Container c806b5f7fde3c0cf227de5b3121205c89ca15e0d3a42aa5bc64156f41944e7b4 is not running
At C:\a\_temp\bf61ed90-b4bd-4edb-8995-26641d616859.ps1:16 char:1

+ Invoke-Command -Session $Session { Param( [HashTable] $Params)

+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    + CategoryInfo          : NotSpecified: (Error response ... is not running:String) [], RemoteException

    + FullyQualifiedErrorId : NativeCommandError

    + PSComputerName        : xxxxxx.westeurope.cloudapp.azure.com
2023-05-26T15:23:36.2743748Z ##[error]PowerShell exited with code '1'.
2023-05-26T15:23:36.2744331Z ##[debug]Processed: ##vso[task.logissue type=error]PowerShell exited with code '1'.
2023-05-26T15:23:36.2856106Z ##[debug]Processed: ##vso[task.complete result=Failed]Error detected
2023-05-26T15:23:36.2889118Z ##[debug]Leaving C:\a\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.220.0\powershell.ps1.
2023-05-26T15:23:36.3133991Z ##[section]Finishing: Configure container

The command is called in a try-catch loop that retries five times, which is why you see the error five times. :)

**Comment by freddydk** [comment]
Didn't notice that you wrote -renewBindings, sorry

When calling restart with renewBindings - it will run one line of code in the container:
```
    if ($renewBindings) {
        Invoke-ScriptInBcContainer -containerName $containerName -ScriptBlock { 
            Set-Content -Path "c:\run\PublicDnsName.txt" -Value ""
        }
    }
```
and for that - yes, the container needs to be running.
I will add a workaround for doing the same if the container is stopped.

**Comment by freddydk** [comment]
BTW - to this comment:

> The container status is "Exited (1073807364) x minutes ago", so it is correct that it doesn't run. But it's never running, before we apply the SSL, and most of the times, it work.

I don't see how Restart-BcContainer with renewBindings would ever have worked with a stopped container???

**Comment by freddydk** [comment]
The fix is shipped in preview

**Comment by epernst** [comment]
Thanks, will test it as soon as possible.

**Comment by freddydk** [comment]
Shipped in 5.0.4
